### PR TITLE
Allow in-place operations on views

### LINKED ISSUE #3313: Support in-place operations on Variable views
Currently, we don't allow in-place operations on views if there are any other live Variables sharing the same data. We should allow in-place operations and compute the gradient correctly. There are a few changes necessary to support this.

Viewing operations include `view`, `narrow`, `select`, `expand`, `[un]squeeze`, and `unfold`. By "a view" I mean a Variable that is a view on another Variable.

1.  Add a field "base" to Variable. Every view has a pointer to a single base Variable. (The base is never a view)
2. In-place operations on views change the `grad_fn` of the base, not of the view.
3. The `grad_fn` on a view may become stale. So views also store an `expected_version`

### In-place operations on views

Every in-place operation on a view inserts three `Functions` to `base.grad_fn`. For example:

```
x = base[2]
x.mul_(y)
```

becomes:

```
Stride(MulBackward(Unstride(<old base.grad_fn>), <y.grad_fn>))
```

Assume, for now, that `base` is contiguous. The Stride op takes the `grad_output`, makes it contiguous, and applies the `stride`, `sizes`, and `storageOffset` of the view (`x`).  This performs the identical viewing operation on `grad_output` that was performed on `base` to get `x`.  For the above example, this is equivalent to:

```
grad_input = grad_output[2]
```

Unstride performs the reverse operation. It creates a `grad_input` with the same shape as `base`, applies the striding of the view `x` to `grad_input`, and copies `grad_output` into that view. For the above example, this is equivalent to:

```
grad_input = zeros_like(base)
grad_input[2] = grad_output
```

`Stride` and `Unstride` are both implemented in terms of a differentiable restride operation that generalizes all of our viewing operations:

```
Tensor restride(Tensor self, IntList sizes, IntList strides, int offset);
```

### Stale grad_fn on views

An in-place operation on a view makes the `grad_fn` of all other views stale. The `compute_next_functions` needs to be updated to handle views: it should re-create the `Function` for a view if it's stale before returning it.

### Non-contiguous base Variables

If base is not contiguous, then we also need the strides/sizes of the base in `restride`:

```
Tensor restride(Tensor self, IntList sizes, IntList strides, IntList base_sizes, IntList base_strides, int offset) {
  Tensor copy = self.type().tensor(base_sizes, base_strides);
  copy.copy_(self);
  return restride(copy, sizes, strides, offset);
}
```

### PR DESCRIPTION
**Pull request by colesbury**
```
Allow in-place operations on views

Adds VariableViewImpl, a subclass of VariableImpl which has a pointer to
the base Variable on which it is a view. In-place operations on views
change the grad_fn of the base.

Fixes #3313
```

### CODE REVIEW

**Comment by soumith** [comment]
retest this please

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
   if (self->cdata.is_executable) {
     self->output_info = new std::vector<output_info_type>();
     self->output_info->reserve(num_outputs);
   }
+
+  // Given an output tensor, find the input Variable with which it shares storage
+  auto get_shared_base = [&shared_pairs, &t2var](PyObject* tensor) -> Variable {
+    auto input_it = t2var.find(tensor);
+    if (input_it != t2var.end()) {
+      // If the output is an input, it's a view on itself
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: The fact that it's in `t2var` inside this function doesn't mean it has to be an input - it can be one of the previous outputs. Not sure if this changes the logic

File: torch/csrc/autograd/python_function.cpp [hunk 2]
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
+    auto it = shared_pairs.find(tensor);
+    if (it != shared_pairs.end()) {
+      // It's explicitly marked as shared via mark_shared_storage
+      return it->second->cdata;
+    }
+    return Variable();
+  };
+
+  // Creates a new Variable from a PyObject* of type THPTensor
+  auto new_variable = [&get_shared_base, is_volatile, cdata](PyObject* output) -> Variable {
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: `cdata` could be captured by reference to avoid a refcount bump

File: torch/csrc/autograd/python_function.cpp [hunk 3]
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
+        // returning the input Variable is not a good option either,
+        // because if someone registers hooks on it, they will fire with grads
+        // from all usages, not only from usages of this output. This is why
+        // we return a view on the input Variable.
+        var = new_variable(output);
       }
     }
+
+    // Wrap new tensors in Variables
+    if (!var.defined()) {
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: this could be an `else` of the `if` above

File: torch/csrc/autograd/python_function.cpp [hunk 4]
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
+    Variable var;
+    auto base = get_shared_base(output);
+    if (base.defined()) {
+      var = make_variable_view(std::move(base), torch::createTensor(output));
+    } else {
+      var = make_variable(torch::createTensor(output));
+    }
+    var.is_volatile() = is_volatile;
+    if (!is_volatile) {
+      var.grad_fn() = cdata;
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: won't this assign a `grad_fn` to things that don't require grad (sth we no longer do for C++ functions)?

File: torch/csrc/autograd/python_function.cpp [hunk 5]
```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
+        _transplant_var(var, cdata, i, is_volatile);
+      } else if (is_leaf) {
+        // An input has been returned, but it wasn't modified. It's better
+        // not to move the Variable, because there are some legitimate cases
+        // where making it non-leaf would break stuff (e.g. broadcast). Also,
+        // returning the input Variable is not a good option either,
+        // because if someone registers hooks on it, they will fire with grads
+        // from all usages, not only from usages of this output. This is why
+        // we return a view on the input Variable.
+        var = new_variable(output);
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: It seems to me that if we get an input Variable that wasn't a leaf **and** wasn't modified, then it won't enter any of those cases. Can you add a comment to show that we thought about this case as well (there was one for it in the old code)?

File: torch/csrc/autograd/variable.h [hunk 6]
```diff
@@ -92,16 +97,20 @@ struct VariableImpl : public at::TensorImpl {
 
 public:
   std::shared_ptr<Function> get_grad_accumulator();
+  virtual std::shared_ptr<Function>& get_grad_fn() {
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: it's a bit annoying that now `grad_fn` lookups are more expensive than for all other attrs

File: torch/csrc/autograd/variable.h [hunk 7]
```diff
@@ -114,6 +123,15 @@ struct VariableImpl : public at::TensorImpl {
...
 };
 
+struct VariableViewImpl : public VariableImpl {
+  VariableViewImpl(Variable base, at::Tensor data);
+
+  virtual std::shared_ptr<Function>& get_grad_fn() override;
+
+  Variable base;
+  int expected_version;
+};
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Can you add a comment briefly describing the view mechanism and mentioning which `VariableImpl` fields might be stale under what conditions?

File: torch/csrc/autograd/variable.cpp [hunk 8]
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+  , expected_version(0) {
+  if (!base.defined()) {
+    throw std::runtime_error("base is undefined");
+  }
+  if (base.is_view()) {
+    base = base.base();
+  }
+  is_view = true;
+  version_counter = base.version_counter();
+  expected_version = version_counter.current_version();
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: I think sth like `attr_version` would describe it better (it's a version that was used to compute the attrs)

File: torch/csrc/autograd/variable.cpp [hunk 9]
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+}
+
+std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+  std::lock_guard<std::mutex> lock(grad_accumulator_lock);
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: can we rename `grad_accumulator_lock` to `var_lock` or sth like this? It seems unrelated to `grad_accumulator` here

File: torch/csrc/autograd/variable.cpp [hunk 10]
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+  if (expected_version != version_counter.current_version()) {
+    auto fn = std::make_shared<generated::AsStridedBackward>();
+    fn->self_geometry = TensorGeometry(base);
+    fn->size = sizes();
+    fn->stride = strides();
+    fn->storage_offset = Variable(this, true).storage_offset();
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: what's happening here? can't you take `data.storage_offset()`?

File: torch/csrc/autograd/variable.cpp [hunk 11]
```diff
@@ -87,6 +89,37 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+    fn->set_flags(Function::flags({ base }));
+    fn->num_inputs = 1;
+    _grad_fn = std::move(fn);
+    expected_version = version_counter.current_version();
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: It seems that there might be a race in here (you read some version, but it changes while you're updating the attrs), although I think it would only happen if someone modifies the base Variable, while you're reading it, which is an error itself. Is that right?

Also, it might be better to get `current_version` once, since it's an atomic load

**Review by apaszke — COMMENTED** [review]
Looks good for the most part, but I think there are a few things that should be changed before this is merged (unless I misunderstood some parts of the code)

**Comment by apaszke** [comment]
BTW do we still need `live_refs` in `VariableVersion`?

**Comment by colesbury** [comment]
I've changed the logic in `_wrap_outputs` some more. It's worth looking over carefully.

**Comment by colesbury** [comment]
I'm still using `live_refs` to prevent in-place operations on views from Python autograd functions. We might be able to change it so that we look at `base` instead.

File: torch/csrc/autograd/python_function.cpp [hunk 12]
```diff
@@ -524,30 +536,23 @@ static void _join_version_counters(THPFunction *self, t2var_type &t2var)
...
+      "mark_shared_storages accepts pairs of tensors, but one of them "
+      "contains %s and %s", THPUtils_typename(t1), THPUtils_typename(t2));
+
+    auto it = t2var.find(t1);
+    THPFunction_assert(it != t2var.end(),
+        "mark_shared_storages only accepts pairs of input "
+        "and output tensors, but argument %d doesn't satify this "
+        "condition", i);
+
+    map.emplace(t2, it->second);
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/python_function.cpp]: Ohhh I didn't notice it's output -> input map, not the other way around. My bad

File: torch/csrc/autograd/variable.cpp [hunk 13]
```diff
@@ -121,6 +122,20 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
   return _grad_fn;
 }
 
+void VariableViewImpl::set_grad_fn(std::shared_ptr<Function> grad_fn) {
+  if (this->_grad_fn && grad_fn) {
+    // If we have a previous grad_fn then this is an in-place modification
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: Views don't have grad functions if they don't require grad, right? So you can have in-place modifications even though you didn't have grad_fn, and the fact that you don't have it doesn't imply that your base still doesn't require grad (it might have been modified in the meantime and you didn't check)

**Review by apaszke — COMMENTED** [review]
In general looks good, but I have a few concerns about correctness. It's really changing the critical parts of autograd and I'm starting to get nervous that it's this large 😕

File: torch/csrc/autograd/python_function.cpp [hunk 14]
```diff
@@ -382,18 +364,19 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
     const t2var_type &shared_pairs,
     PyObject *raw_output, PyObject *outputs, bool is_volatile)
 {
-  auto cdata = is_volatile ? nullptr : THPFunction_asFunction(self);
+  bool is_executable = self->cdata.is_executable && !is_volatile;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_function.cpp]: self->cdata.is_executable should be sufficient

File: torch/csrc/autograd/python_function.cpp [hunk 15]
```diff
@@ -404,80 +387,78 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
...
-      var = new_variable(output);
-    }
+      output_var = (THPVariable*)THPVariable_Wrap(var);
+      if (!output_var) throw python_error();
 
-    THPVariable* output_var = (THPVariable*)THPVariable_Wrap(var);
-    if (!output_var) throw python_error();
+      // We already have the data tensor wrapped as a PyObject*
+      Py_INCREF(output);
+      Py_XDECREF(output_var->data);
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_function.cpp]: Done

File: torch/csrc/autograd/python_function.cpp [hunk 16]
```diff
@@ -404,80 +387,78 @@ static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
...
+      if (auto grad_acc_fn = prev.get()->grad_accumulator.lock()) {
+        auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
+        grad_acc->variable.reset();
+      }
+      return prev;
     }
-    return var;
+    // An input has been returned, but it wasn't modified. Return it as a view
+    // so that we can attach a new grad_fn to the Variable.
+    return make_variable_view(std::move(prev), std::move(data));
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_function.cpp]: 1) Yeah, that should be fixed. test_function_returns_input tests that case

2) Yeah. I think there's a few other simple special case. I'd prefer doing it in a subsequent commit.

File: torch/csrc/autograd/variable.cpp (hunk 13 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: If a view doesn't have a grad_fn, then it's a leaf and we don't currently support in-place operations on leaves.

The logic here worries me too, though. Perhaps we should explicitly have VariableViewImpl::rebase(grad_fn),  which is only called for in-place modifications, instead of trying to determine if it's a modification or new grad_fn internally.

File: torch/csrc/autograd/variable.cpp [hunk 17]
```diff
@@ -121,6 +122,20 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
...
+    if (output_nr != 0 || grad_fn->num_inputs != 1) {
+      throw std::runtime_error("Functions which modify views in-place must return a single Variable");
+    }
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: I forgot to enforce `output_nr` = 0 for views. I'll add those assertions.

**Comment by colesbury** [comment]
Rebased to get TORCH_ASSERT and added two more commits

**Comment by colesbury** [comment]
Here's the diff for the generated VariableType.cpp:

https://gist.github.com/colesbury/bc6835aaef3f755dd063cb484e788701

File: torch/csrc/autograd/functions/tensor.cpp [hunk 18]
```diff
@@ -159,12 +159,18 @@ auto CopySlices::apply(const variable_list& inputs) -> variable_list {
...
   auto res = (*fn)({ grad_slice.clone() });
-  grad_slice.copy_(res[0]);
 
   variable_list grad_inputs(next_functions.size());
-  grad_inputs[0] = result;
-  for (size_t i = 1; i < res.size(); i++) {
-    grad_inputs[i] = std::move(res[i]);
+  for (size_t i = 0; i < res.size(); i++) {
+    if (should_compute_output(i)) {
+      TORCH_ASSERT(res[i].defined());
```
Comment (apaszke) [inline torch/csrc/autograd/functions/tensor.cpp]: I think this assertion is too strict, mostly because `should_compute_output` is really an approximation, that might sometimes have false positives. Right now it's really a check of `is_executable` of `i`th next function. However, it is possible that a function is executable, but its `i`th output was non-differentiable (e.g. it was a max index), and it was used as a target for the loss. This loss will check that its input didn't require grad at runtime, and its backward won't return grad for it.

File: test/test_autograd.py [hunk 19]
```diff
@@ -1542,6 +1542,14 @@ def func(root, b):
         go = Variable(torch.randn(a.size()), requires_grad=True)
         gradgradcheck(func, (a, b), (go,))
 
+    def test_inplace_view5(self):
```
Comment (apaszke) [inline test/test_autograd.py]: Can you please add docstrings to these tests? `test_inplace_view83` won't really tell you what's the essence of the problem and no one will remember that soon. Also, one line docstrings show up in test errors and verbose outputs, which is really nice

File: tools/autograd/gen_variable_type.py [hunk 20]
```diff
@@ -771,19 +768,26 @@ def emit_body(declaration):
...
             env['check_inplace'] = 'if (inplace) check_inplace(input);'
             env['version_counter'] = 'if (inplace) increment_version(input);'
+            env['set_flags'] = SET_FLAGS_INPLACE.substitute(combined, modifies_data='inplace')
+            base_call = 'auto ret = maybe_wrap({}, input, inplace)'.format(base_call)
+        else:
+            env['set_flags'] = SET_FLAGS.substitute(combined)
+            if is_view:
+                base_call = 'auto ret = as_view(static_cast<const Variable&>(self), {})'.format(base_call)
+            else:
+                base_call = 'auto ret = as_variable({})'.format(base_call)
```
Comment (apaszke) [inline tools/autograd/gen_variable_type.py]: Is there any significant change in the logic here, or have you just rearranged things? Is it the addition of `maybe_wrap` (and why is it necessary)?

File: torch/csrc/autograd/variable.h [hunk 21]
```diff
@@ -56,7 +56,10 @@ struct Variable : public at::Tensor {
   inline const std::shared_ptr<Function>& grad_fn() const;
   inline       std::shared_ptr<Function>& grad_fn();
 
+  // Sets the flags and grad_fn ("history") of a new Variable
   inline void set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn);
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: I really thing this would have been much cleaner and safer if we just used the constructor for `set_history` since it's not new Variables only, and `rebase_history` for all other cases

File: torch/csrc/autograd/variable.h [hunk 22]
```diff
@@ -218,14 +220,17 @@ inline std::shared_ptr<Function>& Variable::grad_fn() {
   return get()->get_grad_fn();
 };
 inline void Variable::set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
+  assert(!get()->_grad_fn && "set_history can only be called on new Variables");
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: `TORCH_ASSERT`. This would never get triggered because non-debug builds of Python build extensions with `-DNDEBUG`

File: torch/csrc/autograd/variable.cpp (hunk 13 above)
Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: We do support in-place on leaves if the don't require grad. So how is it finally, is it true that views always have grad_fns or not (my guess is no)?

Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: Yeah, this situations that turn on requires_grad in base seem really dangerous. Any operations on the views will see that they don't require grad, so they won't emit gradient for them, but they should reach the base!

File: tools/autograd/templates/VariableType.h [hunk 23]
```diff
@@ -53,6 +53,7 @@ struct VariableType : public at::Type {
   std::tuple<Variable, Variable> as_variable(std::tuple<Tensor, Tensor> tensor) const;
   std::tuple<Variable, Variable, Variable> as_variable(std::tuple<Tensor, Tensor, Tensor> tensor) const;
   std::vector<Variable> as_variable(TensorList tensor) const;
+  Variable maybe_wrap(Tensor data, const Variable & self, bool inplace) const;
```
Comment (apaszke) [inline tools/autograd/templates/VariableType.h]: I can't see a reason why would this need to be a `VariableType` method (it doesn't even use `this`) and not a static helper in the file. I think it would be clearer that way

File: torch/csrc/autograd/variable.cpp [hunk 24]
```diff
@@ -122,18 +124,22 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
...
+  if (!grad_fn) {
+    TORCH_ASSERTM(!requires_grad, "Can't set null grad_fn on view with requires_grad=True");
+    TORCH_ASSERTM(!base.requires_grad(), "base.requires_grad does not match view.requires_grad");
+    return;
+  }
+
+  TORCH_ASSERTM(requires_grad, "Can't set grad_fn on view with requires_grad=False");
+  TORCH_ASSERT(output_nr == 0);
+  if (grad_fn->num_inputs != 1) {
+    throw std::runtime_error("Functions which modify views in-place must return a single Variable");
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: Is that right? I thought the only limitation was that views have to be the first output. Or is it because `CopySlices` would have to handle that?

File: torch/csrc/autograd/variable.h [hunk 25]
```diff
@@ -42,9 +51,16 @@ struct Variable : public at::Tensor {
   inline const Variable & grad() const;
   inline       Variable & grad();
 
+  inline bool is_leaf() const;
+
   inline const std::shared_ptr<Function>& grad_fn() const;
   inline       std::shared_ptr<Function>& grad_fn();
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Can we make it always return a const shared pointer? I think all modifications to `grad_fn` should be through `set_history` and `rebase_history`, right?

File: torch/csrc/autograd/python_function.cpp [hunk 26]
```diff
@@ -382,87 +356,101 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
 // mappings for output tensors as well.
 static void _wrap_outputs(THPFunction *self, t2var_type &t2var,
-    std::unordered_set<PyObject *> &dirty_inputs, PyObject *raw_output,
-    PyObject *outputs, bool is_volatile)
+    std::unordered_set<PyObject *> &dirty_inputs,
+    const t2var_type &shared_pairs,
+    PyObject *raw_output, PyObject *outputs, bool is_volatile)
 {
-  auto cdata = is_volatile ? nullptr : THPFunction_asFunction(self);
+  bool is_executable = self->cdata.is_executable && !is_volatile;
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: Can you remove `!is_volatile`? The first condition should be enough (as discussed in the previous review)

**Review by apaszke — COMMENTED** [review]
Looks good, but we should discuss the situation when `requires_grad` of `base` changes

File: tools/autograd/templates/VariableType.h (hunk 23 above)
Comment (colesbury) [reply tools/autograd/templates/VariableType.h]: It calls `as_variable` which is on `VariableType`. `as_variable` has to be on VariableType because the overload for `Scalar`uses `this->baseType`. C++ doesn't allow a mix of overloads (`as_variable`) that are methods and functions.

Comment (colesbury) [reply tools/autograd/templates/VariableType.h]: This can be simplified once ATen only returns "Tensor" instead of "Scalar"

File: torch/csrc/autograd/variable.cpp (hunk 24 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: Yeah, `CopySlices` would have to handle that, which is difficult and doesn't seem worth it for now at least.

File: torch/csrc/autograd/variable.cpp [hunk 27]
```diff
@@ -108,6 +108,14 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_)
 
 std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
   std::lock_guard<std::mutex> lock(mutex);
+  if (base.requires_grad() && !requires_grad) {
+    // TODO: See test_inplace_view6. It would be good to support this operation
+    // but that might require sharing requires_grad between the base and the view
+    throw std::runtime_error(
+        "requires_grad is False and base.requires_grad is True. Cannot use "
+        "this view in a differentiable operation. Re-create the view from the "
+        "base Variable.");
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: It might be a bit more helpful to say "re-create the view form its base Variable after the last in-place modification".

File: torch/csrc/autograd/python_variable.cpp [hunk 28]
```diff
@@ -391,6 +431,7 @@ static struct PyGetSetDef THPVariable_properties[] = {
   {"data", (getter)THPVariable_get_data, (setter)THPVariable_set_data, NULL, NULL},
   {"_grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, NULL, NULL}, // only for legacy reasons
   {"grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, NULL, NULL},
+  {"_base", (getter)THPVariable_get_base, NULL, NULL, NULL}, // only for legacy reasons
```
Comment (apaszke) [inline torch/csrc/autograd/python_variable.cpp]: what legacy reasons?

**Review by apaszke — COMMENTED** [review]
Last commits look good to me, but please add more comments to the tests to clearly state what you're checking. I need one more day to look go over `VariableViewImpl` again and ensure that there are no more weird edge cases.

File: torch/csrc/autograd/python_variable.cpp (hunk 28 above)
Comment (colesbury) [reply torch/csrc/autograd/python_variable.cpp]: oops - copy pasta

File: torch/csrc/autograd/variable.cpp [hunk 29]
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+  , attr_version(0) {
+  if (!base.defined()) {
+    throw std::runtime_error("base is undefined");
+  }
+  if (base.is_view()) {
+    base = base.base();
+  }
+  is_view = true;
+  version_counter = base.version_counter();
+  attr_version = version_counter.current_version();
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: I think it would be safer to set `attr_version` to an invalid value like -1. You didn't set up a `grad_fn` in here, but it might be necessary, so we better enter the slow path of `get_grad_fn()` next time

File: torch/csrc/autograd/variable.cpp [hunk 30]
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+}
+
+std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+  std::lock_guard<std::mutex> lock(mutex);
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: It seems to me that it would be enough to acquire the mutex in the slow path.

File: torch/csrc/autograd/variable.cpp [hunk 31]
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+
+void VariableViewImpl::rebase_grad_fn(std::shared_ptr<Function> grad_fn) {
+  if (!grad_fn) {
+    TORCH_ASSERTM(!requires_grad, "Can't set null grad_fn on view with requires_grad=True");
+    TORCH_ASSERTM(!base.requires_grad(), "base.requires_grad does not match view.requires_grad");
+    return;
+  }
+
+  TORCH_ASSERTM(requires_grad, "Can't set grad_fn on view with requires_grad=False");
+  TORCH_ASSERT(output_nr == 0);
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: Can you move this assert to the beginning, so it is always checked?

Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: Also, shouldn't we assert that `base.requires_grad()` matches the value of this function?

File: torch/csrc/autograd/variable.cpp [hunk 32]
```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
...
+  if (grad_fn->num_inputs != 1) {
+    throw std::runtime_error("Functions which modify views in-place must return a single Variable");
+  }
+  auto copySlices = std::make_shared<CopySlices>(base, TensorGeometry(data), std::move(grad_fn));
+  base.output_nr() = 0;
+  base.get()->_grad_fn = std::move(copySlices);
+  base.requires_grad() = true;
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: This line depends on the check that in-place functions return only a single value. Can you please add a warning or sth?

File: torch/csrc/autograd/variable.h [hunk 33]
```diff
@@ -155,12 +209,26 @@ inline Variable & Variable::grad() {
...
+
 inline const std::shared_ptr<Function>& Variable::grad_fn() const {
-  return get()->grad_fn;
-};
-inline std::shared_ptr<Function>& Variable::grad_fn() {
-  return get()->grad_fn;
+  return get()->get_grad_fn();
 };
+inline void Variable::set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
+  assert(!get()->_grad_fn && "set_history can only be called on new Variables");
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Use `TORCH_ASSERT`. Plus, this doesn't really imply that it's a new Variable (it might be an old Variable that only now starts requiring grad).

Comment (apaszke) [reply torch/csrc/autograd/variable.h]: Also, can you add an assertion that `grad_fn` is null if `flags.requires_grad` is false? Same below

File: torch/csrc/autograd/python_function.cpp [hunk 34]
```diff
@@ -382,87 +356,102 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
...
-        if (dirty_inputs.count(output) > 0) {
-          _transplant_var(*input_var_.get(), cdata, i, is_volatile);
-        }
+      // Wrap the output in a Variable
+      bool is_modified = dirty_inputs.count(output) > 0;
+      Variable var = wrap_output(
+          torch::createTensor(output),
+          get_shared_base(output),
+          is_modified);
+      if (is_modified) {
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: Correctness of these lines depends on the code of `wrap_output` which is a bit far right now. Can you add a comment saying that it's safe to just call `set_history`, because the only case where we return the original variable is when it was an input and modified, so it will always enter the first case?

**Review by apaszke — COMMENTED** [review]
Alright, I think I'm ok with merging this once you address these comments. Please please please please add the test descriptions as well, I have no idea what are they checking.

File: torch/csrc/autograd/variable.cpp (hunk 30 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: The read on `_grad_fn` should be guarded by the mutex, since another call to `get_grad_fn` could modify it.

File: torch/csrc/autograd/variable.cpp (hunk 32 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: There's an assertion for that immediately above this.

Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: I know, but it's not immediately clear that this line depends on the assertion. Adding a tiny comment saying that it is like this won't hurt, and we won't miss it in the future for sure

File: torch/csrc/autograd/variable.h (hunk 33 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.h]: We still create grad_fn for Python-implemented functions that don't require grad. (I'm going to change that in a subsequent PR).

File: torch/csrc/autograd/python_function.cpp (hunk 34 above)
Comment (colesbury) [reply torch/csrc/autograd/python_function.cpp]: I've moved most of `set_history` into `make_variable` and moved the rebase call up to wrap_output.

File: torch/csrc/autograd/variable.cpp (hunk 29 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: I've changed the constructor so it takes a grad_fn

File: torch/csrc/autograd/variable.cpp (hunk 32 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: What line are you referring to? The `base.output_nr() = 0` doesn't depend on that assertion.

**Comment by colesbury** [comment]
I've removed the calls to `set_history` in-favor of passing those arguments to Variable factory functions. The generated VariableType.cpp still effectively calls "set_history", but it's inlined into the calling code. In a subsequent PR, I'll clean up VariableType.cpp. I don't want to do it here because that change is probably substantial and this isn't a correctness issue.

File: torch/csrc/autograd/variable.cpp [hunk 35]
```diff
@@ -91,8 +86,9 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
   return result;
 }
 
-VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_)
-  : VariableImpl(std::move(data_))
+VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, VarFlags flags,
+                                   int output_nr, std::shared_ptr<Function> grad_fn)
+  : VariableImpl(std::move(data_), flags, output_nr, std::move(grad_fn))
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: nit: we get `output_nr` now, so add `TORCH_ASSERT(output_nr == 0)`

File: torch/csrc/autograd/variable.h [hunk 36]
```diff
@@ -164,20 +164,20 @@ struct VariableViewImpl : public VariableImpl {
...
 }
 
-inline Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn) {
-  return Variable(new VariableImpl(std::move(data), std::move(grad_fn)), false);
-}
+Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn);
 
 inline Variable make_variable(at::Tensor data, bool requires_grad, bool is_volatile=false) {
-  return Variable(new VariableImpl(std::move(data), requires_grad, is_volatile), false);
+  return make_variable(std::move(data), VarFlags(requires_grad, is_volatile));
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: this overload is a bit weird now, as everything else uses `VarFlags`. but I don't mind having it

**Review by apaszke — APPROVED** [review]
LGTM

**Comment by apaszke** [comment]
Actually you haven't added a check for the case when we're rebasing a view that requires grad (we should check that base still requires grad as well)

File: torch/csrc/autograd/variable.h (hunk 36 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.h]: Yeah, we should clean this up soon
